windows/hid.c

```diff
@@ -358,28 +387,98 @@ HID_API_EXPORT const char* HID_API_CALL hid_version_str(void)
 	return HID_API_VERSION_STR;
 }
 
+static void hid_internal_hotplug_init()
+{
+	if (!hid_hotplug_context.critical_section_ready) {
+		InitializeCriticalSection(&hid_hotplug_context.critical_section);
+		hid_hotplug_context.critical_section_ready = 1;
+	}
+}
+
 int HID_API_EXPORT hid_init(void)
 {
 	register_global_error(NULL);
+
 #ifndef HIDAPI_USE_DDK
 	if (!hidapi_initialized) {
 		if (lookup_functions() < 0) {
 			register_global_winapi_error(L"resolve DLL functions");
 			return -1;
-		}
+        }
 		hidapi_initialized = TRUE;
 	}
 #endif
+
 	return 0;
 }
 
+static void hid_internal_hotplug_cleanup()
+{
+	/* Unregister the HID device connection notification when removing the last callback */
+	/* This function is always called inside a locked mutex */
+	if (hid_hotplug_context.hotplug_cbs != NULL) {
+		return;
+	}
+
+	if (hid_hotplug_context.devs) {
+		/* Cleanup connected device list */
+		hid_free_enumeration(hid_hotplug_context.devs);
+		hid_hotplug_context.devs = NULL;
+	}
+
+	if (hid_hotplug_context.notify_handle) {
+		if (CM_Unregister_Notification(hid_hotplug_context.notify_handle) != CR_SUCCESS) {
+			register_global_error(
+				L"CM_Unregister_Notification failed for Hotplug notification");
+			return;
```

Comment:
even if we've failed here - we probably still need to clear the value of `notify_handle` so we do not attempt to Unregister it once more